quickannotator/db/crud/tile.py

```diff
@@ -0,0 +1,284 @@
+import numpy as np
+from shapely.affinity import scale
+from shapely.geometry import Polygon, shape
+from sqlalchemy.dialects.postgresql import insert as pg_insert
+from sqlalchemy.dialects.sqlite import insert as sqlite_insert
+from sqlalchemy import case, update
+from quickannotator.api.v1.utils.coordinate_space import base_to_work_scaling_factor, get_tilespace
+from quickannotator.constants import MASK_CLASS_ID, MASK_DILATION, TileStatus
+import quickannotator.constants as constants
+from datetime import datetime, timedelta
+from quickannotator.db import db_session
+from quickannotator.db.crud.annotation import get_annotation_query
+from quickannotator.db.crud.annotation_class import get_annotation_class_by_id
+from quickannotator.db.crud.image import get_image_by_id
+import quickannotator.db.models as db_models
+from quickannotator.db.models import get_model_column_names
+import geojson
+import cv2
+
+from typing import List
+from abc import ABC
+
+from quickannotator.db.utils import build_annotation_table_name, create_dynamic_model
+
+
+class TileStore(ABC):   # Only an ABC to prevent instantiation
+    def __init__(self, insert_method=None):
+        self.insert_method = insert_method
+
+    def _upsert_tiles(self, image_id, annotation_class_id, tile_ids, insert_fields, update_fields):
+        """
+        Inserts new tiles or updates existing tiles in the database.
+        This function attempts to insert new tiles with the given annotation_class_id, image_id, and tile_ids.
+        If tiles with the same annotation_class_id, image_id, and tile_ids already exist, it updates the relevant fields.
+
+        Args:
+            image_id (int): The ID of the image.
+            annotation_class_id (int): The ID of the annotation class.
+            tile_ids (List[int]): The IDs of the tiles.
+            insert_fields (dict): Additional fields to insert into the tile.
+            update_fields (dict): Additional fields to update in the tile.
+        Returns:
+            List[models.Tile]: The tiles that were inserted or updated.
+        """
+        if len(tile_ids) == 0:
+            return []
+
+        stmt = self.insert_method(db_models.Tile).values(
+            [
+                {
+                    'annotation_class_id': annotation_class_id,
+                    'image_id': image_id,
+                    'tile_id': tile_id,
+                    **insert_fields
+                }
+                for tile_id in tile_ids
+            ]
+        ).on_conflict_do_update(
+            index_elements=['annotation_class_id', 'image_id', 'tile_id'],
+            set_=update_fields
```

Comment:
does set have an underscore? set_?

Reply:
Yes, here's the postgis docs, and the interface is identical for the spatialite dialect insert method

https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.on_conflict_do_update.params.set_

Reply:
Looked it up and it's because "set" is a reserved keyword in python.

Reply:
yea - that makes sense....however if i were in their position i would likely think of another word to use instead of set, to avoid both hitting into a reserve word AND this weird underscore pattern (which i don't think i've seen anywhere else?)